### PR TITLE
UCP/RNDV: Add 2-stage ppln protocol for protocols v2

### PR DESCRIPTION
**Pull request by brminich**
## What
Complete adding 2-stage ppln protocol for protocols v2 (#8526)

### CODE REVIEW

File: src/ucp/rndv/proto_rndv.c [hunk 1]
```diff
@@ -60,11 +63,13 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
             continue;
         }
 
-        /* Check the memory domain requires remote key, and capable of
-         * registering the memory type
-         */
-        if (!(md_attr->flags & UCT_MD_FLAG_NEED_RKEY) ||
-            !(context->reg_md_map[params->mem_info.type] & UCS_BIT(md_index))) {
+        if (!(params->alloc_md_index == md_index) &&
```
Comment (yosefe) [inline src/ucp/rndv/proto_rndv.c:66]: can you document this condition? is it still needed?

File: src/ucp/rndv/proto_rndv.c [hunk 2]
```diff
@@ -580,7 +586,7 @@ ucp_proto_rndv_bulk_init(const ucp_proto_multi_init_params_t *init_params,
     return status;
 }
 
-static size_t ucp_proto_rndv_ats_pack_ack(void *dest, void *arg)
+size_t ucp_proto_rndv_common_pack_ack(void *dest, void *arg)
```
Comment (yosefe) [inline src/ucp/rndv/proto_rndv.c:589]: ucp_proto_rndv_pack_ack

File: src/ucp/rndv/proto_rndv.c [hunk 3]
```diff
@@ -672,7 +678,10 @@ ucp_proto_rndv_send_reply(ucp_worker_h worker, ucp_request_t *req,
         goto err_destroy_rkey;
     }
 
-    req->send.rndv.rkey = rkey;
+    req->send.rndv.rkey        = rkey;
+    /* Caching rkey_buffer pointer for later unpacking of shm keys in
+     * rkey_ptr mtype ppln protocol. */
+    req->send.rndv.rkey_buffer = (void*)rkey_buffer;
```
Comment (yosefe) [inline src/ucp/rndv/proto_rndv.c:684]: can probably remove `(void*)`

File: src/ucp/rndv/rndv_mtype.inl [hunk 4]
```diff
@@ -50,8 +50,10 @@ ucp_proto_rndv_mtype_request_init(ucp_request_t *req)
 }
 
 static UCS_F_ALWAYS_INLINE uct_mem_h
-ucp_proto_rndv_mtype_get_memh(ucp_mem_desc_t *mdesc, ucp_rsc_index_t memh_index)
+ucp_proto_rndv_mtype_get_memh(ucp_request_t *req, ucp_md_index_t memh_index)
```
Comment (yosefe) [inline src/ucp/rndv/rndv_mtype.inl:53]: rename memh_index param to md_index

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 5]
```diff
@@ -176,10 +185,171 @@ ucp_proto_t ucp_rndv_rkey_ptr_proto = {
...
+
+    return status;
+}
+
+static ucs_status_t ucp_proto_rndv_rkey_ptr_mtype_completion(ucp_request_t *req)
+{
+    ucp_trace_req(req, "ucp_proto_rndv_rkey_ptr_mtype_completion");
+    ucp_proto_rndv_rkey_destroy(req);
+    ucp_proto_request_zcopy_complete(req, UCS_OK);
+
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:270]: minor: remove spaceline

File: src/ucp/rndv/rndv_rkey_ptr.c [hunk 6]
```diff
@@ -176,10 +185,171 @@ ucp_proto_t ucp_rndv_rkey_ptr_proto = {
...
+    ucs_assert(rkey_buffer != NULL);
+
+    req->send.rndv.rkey_buffer = NULL;
+
+    ppln_data = ucp_ep_peer_mem_get(context, req->send.ep, remote_address,
+                                    req->send.state.dt_iter.length,
+                                    rkey_buffer, local_mem_type,
+                                    rpriv->dst_md_index);
+    if (ppln_data->rkey == NULL) {
+        ucs_error("unreachable rkey for md_index %u", rpriv->dst_md_index);
```
Comment (yosefe) [inline src/ucp/rndv/rndv_rkey_ptr.c:307]: print more info

File: src/ucp/rndv/proto_rndv.c (hunk 2 above)
Comment (brminich) [reply src/ucp/rndv/proto_rndv.c:589]: `ucp_proto_rndv_pack_ack` already exists

File: src/ucp/rndv/rndv_rkey_ptr.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/rndv/rndv_rkey_ptr.c:306]: seems unresolved

**Review by yosefe — COMMENTED** [review]
https://github.com/openucx/ucx/pull/8668/files/041c37e73fa4342025a24f2f952a05fc6d907f63#diff-a87749a9f956c2a0866fdb72d3144df743f189b2dcf47238fbbf57c0caeca261

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**